packages/field-base/src/text-field-mixin.js

```diff
@@ -0,0 +1,124 @@
+/**
+ * @license
+ * Copyright (c) 2021 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+import { Debouncer } from '@polymer/polymer/lib/utils/debounce.js';
+import { timeOut } from '@polymer/polymer/lib/utils/async.js';
+import { dedupingMixin } from '@polymer/polymer/lib/utils/mixin.js';
+import { InputFieldMixin } from './input-field-mixin.js';
+
+const TextFieldMixinImplementation = (superclass) =>
+  class TextFieldMixinClass extends InputFieldMixin(superclass) {
+    static get properties() {
+      return {
+        /**
+         * Maximum number of characters (in Unicode code points) that the user can enter.
+         */
+        maxlength: {
+          type: Number
+        },
+
+        /**
+         * Minimum number of characters (in Unicode code points) that the user can enter.
+         */
+        minlength: {
+          type: Number
+        },
+
+        /**
+         * A regular expression that the value is checked against.
+         * The pattern must match the entire value, not just some subset.
+         */
+        pattern: {
+          type: String
+        },
+
+        /**
+         * When set to true, user is prevented from typing a value that
+         * conflicts with the given `pattern`.
+         * @attr {boolean} prevent-invalid-input
+         */
+        preventInvalidInput: {
+          type: Boolean
+        }
+      };
+    }
+
+    static get hostProps() {
+      return [...super.hostProps, 'pattern', 'maxlength', 'minlength'];
+    }
+
+    /** @protected */
+    ready() {
+      super.ready();
+
+      this._createConstraintsObserver();
+    }
+
+    /** @protected */
+    _createConstraintsObserver() {
+      // This complex observer needs to be added dynamically here (instead of defining it above in the `get observers()`)
+      // so that it runs after complex observers of inheriting classes. Otherwise e.g. `_stepOrMinChanged()` observer of
+      // vaadin-number-field would run after this and the `min` and `step` properties would not yet be propagated to
+      // the `inputElement` when this runs.
+      this._createMethodObserver('_constraintsChanged(required, minlength, maxlength, pattern)');
+    }
+
+    /**
+     * @param {boolean | undefined} required
+     * @param {number | undefined} minlength
+     * @param {number | undefined} maxlength
+     * @param {string | undefined} maxlength
+     * @protected
+     */
+    _constraintsChanged(required, minlength, maxlength, pattern) {
+      if (!this.invalid) {
+        return;
+      }
```

Comment:
Not sure I understand this one. Let's assume an empty input with no constraints. Then I set a constraint like minLength=3 programmatically. Then I don't want to immediatly show that the input is invalid?

I can only assume that the validation should first trigger when the user interacts and blurs the input? I guess there can easily be a case where you create an input and only afterwards set the constraint, which would immediatly show it is invalid, even though the user did not have a chance to interact with it. Just voicing my thoughts here. But maybe we should document this behavior somehow for future readers.

Reply:
Originally it was added by https://github.com/vaadin/vaadin-text-field/pull/401 but I'm not sure why this check was there. Removed this part altogether, I could not find any use case where it would be needed.

Reply:
Restored the code and added a missing test after discussing this in person.